symbolic-base/src/ZkFold/FFI/Rust/RustBLS.hs

```diff
@@ -0,0 +1,496 @@
+  polyVecZero n = h2r $ polyVecZero n
+
+  polyVecLagrange :: forall size . (KnownNat size) => Natural -> Natural -> Fr -> RustPolyVec Fr size
+  polyVecLagrange n i omega = h2r $ polyVecLagrange n i (r2h omega)
+
+  polyVecInLagrangeBasis :: forall n size . (KnownNat n, KnownNat size) => Fr -> RustPolyVec Fr n -> RustPolyVec Fr size
+  polyVecInLagrangeBasis omega pv = h2r $ polyVecInLagrangeBasis (r2h omega) (r2h pv)
+
+  polyVecGrandProduct :: forall size . (KnownNat size) => RustPolyVec Fr size -> RustPolyVec Fr size -> RustPolyVec Fr size -> Fr -> Fr -> RustPolyVec Fr size
+  polyVecGrandProduct a b sigma beta gamma = h2r $ polyVecGrandProduct (r2h a) (r2h b) (r2h sigma) (r2h beta) (r2h gamma)
+
+  polyVecDiv :: forall size . (KnownNat size) => RustPolyVec Fr size -> RustPolyVec Fr size -> RustPolyVec Fr size
+  polyVecDiv l r = unsafePerformIO runFFT
+    where
+      runFFT :: IO (RustPolyVec Fr size)
+      runFFT = do
+        let valueSize = (fromInteger $ naturalToInteger $ value @size)
+        out <- callocForeignPtrBytes @CChar (scalarSize P.* (fromInteger $ naturalToInteger $ value @size))
+
+        runRustFunctionBinary rsDivFFT
+          (passPolyVec l)
+          (passPolyVec r)
+          (out, valueSize P.* scalarSize)
+        return $ RustPV (RData out)
+
+  castPolyVec :: forall size size' . (KnownNat size, KnownNat size') => RustPolyVec Fr size -> RustPolyVec Fr size'
+  castPolyVec pv = h2r $ castPolyVec (r2h pv)
+
+instance (KnownNat size) => Scale Natural (RustPolyVec Fr size) where
+  scale c pv = h2r $ (scale c (r2h pv) :: PolyVec EC.Fr size)
+
+instance (KnownNat size) => Scale Integer (RustPolyVec Fr size) where
+  scale c pv = h2r $ (scale c (r2h pv) :: PolyVec EC.Fr size)
+
+instance (KnownNat size) => FromConstant Natural (RustPolyVec Fr size) where
+  fromConstant n = h2r $ fromConstant n
+
+instance (KnownNat size) => FromConstant Integer (RustPolyVec Fr size) where
+  fromConstant n = h2r $ fromConstant n
+
+instance (KnownNat size) => AdditiveSemigroup (RustPolyVec Fr size) where
+  l + r = unsafePerformIO runSum
+    where
+      runSum :: IO (RustPolyVec Fr size)
+      runSum = do
+        let valueSize = (fromInteger $ naturalToInteger $ value @size)
+        out <- callocForeignPtrBytes @CChar (scalarSize P.* (fromInteger $ naturalToInteger $ value @size))
+
+        runRustFunctionBinary rsAdd
+          (passPolyVec l)
+          (passPolyVec r)
+          (out, valueSize P.* scalarSize)
+        return $ RustPV (RData out)
+
+instance (KnownNat size) => AdditiveMonoid (RustPolyVec Fr size) where
+  zero = h2r $ (zero :: PolyVec EC.Fr size)
+
+instance (KnownNat size) => AdditiveGroup (RustPolyVec Fr size) where
+  negate pv = h2r $ negate (r2h pv)
+
+instance (KnownNat size) => Exponent (RustPolyVec Fr size) Natural where
+  pv ^ n = h2r $ (^) (r2h pv) n
+
+instance {-# OVERLAPPING #-} (KnownNat size) => Scale (RustPolyVec Fr size) (RustPolyVec Fr size)
+
+-- TODO (Issue #18): check for overflow
```

Comment:
Issue number is from `zkfold-prover` repo, I guess we need to move the issue to `symbolic-base` as well

Reply:
I think this is a copypaste from Symbolic. If you take a look, we have a lot of "Hide this constructor" issues